### ISSUE TITLE
Some ntpath/posixpath functions unusable on foreign OS

### ISSUE THREAD

**Issue by barneygale** [python/cpython#119671]
The [`os.path` docs intro](https://docs.python.org/3/library/os.path.html) includes:

> Note: Since different operating systems have different path name conventions, there are several versions of this module in the standard library. The `os.path` module is always the path module suitable for the operating system Python is running on, and therefore usable for local paths. However, you can also import and use the individual modules if you want to manipulate a path that is *always* in one of the different formats. They all have the same interface:
>
> - `posixpath` for UNIX-style paths
> - `ntpath` for Windows paths

But we don't explain that some of the functions in `ntpath` and `posixpath` aren't suitable for usage on a "foreign" OS, e.g. that using `ntpath.realpath()` from POSIX is a bad idea.

These functions are safe because they do purely lexical work:

- `basename`
- `commonpath`
- `commonprefix`
- `dirname`
- `isabs`
- `join`
- `normcase`
- `normpath`
- `split`
- `splitdrive`
- `splitroot`
- `splitext`

These functions are probably unsafe because they rely on details of the host OS (e.g. via the `os` module):

- `abspath`
- `exists`
- `lexists`
- `expanduser`
- `expandvars`
- `getatime`
- `getmtime`
- `getctime`
- `getsize`
- `isfile`
- `isdir`
- `isjunction`
- `islink`
- `ismount`
- `isdevdrive`
- `realpath`
- `relpath`
- `samefile`
- `sameopenfile`
- `samestat`
- `supports_unicode_filenames`

### Linked PRs
* gh-119772

**Comment by barneygale** [comment]
I'm not sure if/how to solve this.

It could be considered a docs bug - perhaps we need to adjust the `os.path` docs to only recommend `ntpath` and `posixpath` in certain circumstances, and document the safe interface.

Alternatively we could remove the docs mention of the `ntpath` and `posixpath` modules (making them quasi-private again), and instead add `os.PosixPathParser` and `os.WindowsPathParser` classes that expose only the safe methods. There's something similar in the pathlib ABCs already:

https://github.com/python/cpython/blob/6ec371223dff4da7719039e271f35a16a5b861c6/Lib/pathlib/_abc.py#L35-L84

**Comment by nineteendo** [comment]
Also, it's worth noting that there's no problem with `relpath(abs1, abs2)`.

**Comment by nineteendo** [comment]
Could we maybe move the safe functions to `ntpath.abstract` & `posixpath.abstract`, re-exporting them in `ntpath.__init__.py` & `posixpath.__init__.py`?

**Comment by barneygale** [comment]
> Also, it's worth noting that there's no problem with `relpath(abs1, abs2)`.

It can call `abspath()` which can call `os.getcwd()`

**Comment by nineteendo** [comment]
Could we discuss adding a submodule on Discourse? I was asked to not create new idea threads until 2025.

**Comment by barneygale** [comment]
TBH I consider this a pretty low-priority issue and not worth raising on discourse. It's been this way for a long time. If/when other core devs encounter this issue, they might search the bug tracker, arrive here, and give their tuppence.

**Comment by zooba** [comment]
I don't think this is worth adding submodules. Fallback implementations or `raise NotImplementedError` for those that are _really_ unusable ought to be enough. Availability should already be documented, but we can fix those if necessary (and static analysis tools could/should use that information to provide warnings that we don't).

We've got little to no chance of moving any existing functions into new modules, so no compatibility issues are resolved. But I think we can justify replacing unexplained OSError/AttributeErrors with NotImplementedError (or a `NotSupportedError` if we have one?) across a set of functions.

**Comment by nineteendo** [comment]
Note that even this conservative check doesn't work for `relpath()` as `zipfile.Path.relative_to()` relies on this handling the garbage on Windows, which I **really** don't like.

```diff
diff --git a/Lib/posixpath.py b/Lib/posixpath.py
index fccca4e066b..dd431c4dbf7 100644
--- a/Lib/posixpath.py
+++ b/Lib/posixpath.py
@@ -516,8 +516,13 @@ def relpath(path, start=None):
         start = os.fspath(start)
 
     try:
-        start_tail = abspath(start).lstrip(sep)
-        path_tail = abspath(path).lstrip(sep)
+        start_abs = abspath(start)
+        path_abs = abspath(path)
+        if not isabs(start_abs) or not isabs(path_abs):
+            raise ValueError("paths are not absolute")
+
+        start_tail = start_abs.lstrip(sep)
+        path_tail = path_abs.lstrip(sep)
         start_list = start_tail.split(sep) if start_tail else []
         path_list = path_tail.split(sep) if path_tail else []
         # Work out how much of the filepath is shared by start and path.
```

At least the submodules could make this work without breaking anyone's code. And all safe functions can be refactored to them, making the code easier to maintain. Other approaches actually seem more complicated to me.
